api/src/test/java/org/bahmni/module/communication/service/impl/CommunicationServiceImplTest.java

```diff
@@ -0,0 +1,214 @@
+        verify(mailTransport).sendMessage(mimeMessageCaptor.capture(), any());
+        int bodyPartCount = ((MimeMultipart) mimeMessageCaptor.getValue().getContent()).getCount();
+        assertEquals(bodyPartCount, 1); //only contains email body
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenItFailsToSendEmail() throws MessagingException {
+        doThrow(new MessagingException()).when(mailTransport).sendMessage(any(), any());
+        assertThrows(RuntimeException.class, () -> communicationService.sendEmail(new MailContent()));
+    }
+
+    @Test
+    public void shouldSendSMS() throws IOException {
+        String phoneNumber = "1234567890";
+        String message = "This is a test message";
+        SMSRequest expectedSmsRequestObject = new SMSRequest();
+        expectedSmsRequestObject.setPhoneNumber(phoneNumber);
+        expectedSmsRequestObject.setMessage(message);
+        String expectedSmsRequest = new ObjectMapper().writeValueAsString(expectedSmsRequestObject);
+        String tokenFilePath = new File(OpenmrsUtil.getApplicationDataDirectory() + "/sms-token/sms-communications-token.txt").getAbsolutePath();
+        String expectedToken = "expected_token";
+        String smsUrl = "http://localhost:25/sms";
+
+        when(messagingUtility.getSMSTokenFromTokenFile(tokenFilePath)).thenReturn(expectedToken);
+        when(administrationService.getGlobalProperty("bahmni.sms.url")).thenReturn(smsUrl);
+        when(messageSourceService.getMessage(smsUrl, null, new Locale("en"))).thenReturn(smsUrl);
+
+        communicationService.sendSMS(phoneNumber, message);
+
+        ArgumentCaptor<HttpPost> captorRequest = ArgumentCaptor.forClass(HttpPost.class);
+        verify(httpClient).execute(captorRequest.capture());
+        verify(httpClient).close();
+        HttpPost actualRequest = captorRequest.getValue();
+        assertEquals(smsUrl, actualRequest.getURI().toString());
+        assertEquals("content-type: application/json", actualRequest.getAllHeaders()[0].toString());
+        assertEquals("Authorization: Bearer " + expectedToken, actualRequest.getAllHeaders()[1].toString());
+        assertEquals(expectedSmsRequest, EntityUtils.toString(actualRequest.getEntity()));
+    }
+
+    @Test
+    public void shouldAbortRequestWhenSMSUrlIsEmpty() throws IOException {
+        when(administrationService.getGlobalProperty("bahmni.sms.url")).thenReturn("");
+
+        communicationService.sendSMS("1234567890", "This is a test message");
+
+        verify(httpClient, never()).execute(any());
+    }
+
+    @Test
+    public void shouldAbortRequestWhenSMSUrlIsNull() throws IOException {
+        when(administrationService.getGlobalProperty("bahmni.sms.url")).thenReturn(null);
+
+        communicationService.sendSMS("1234567890", "This is a test message");
+
+        verify(httpClient, never()).execute(any());
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenTokenNotFound() {
```

Comment:
Should we add a validation for errorMessage, for easy debugging of this issue?

Reply:
Yes, will do that!